File: telemetry/aitf/sdk/rust/src/exporters/cef_syslog.rs
//! CEF (Common Event Format) syslog exporter.
//!
//! Renders [`AIBaseEvent`] values as CEF syslog message strings for SIEMs that
//! do not support OCSF natively (ArcSight, QRadar, LogRhythm, Elastic Security,
//! etc.). Ported from the Go `exporters/cef_syslog.go`; the class-UID name map,
//! the OCSF-severity -> CEF-severity map, header/extension layout, and value
//! sanitization match that implementation.

use serde_json::Value;

use crate::ocsf::AIBaseEvent;

/// Maps OCSF `severity_id` to a CEF severity (0..10).
fn ocsf_to_cef_severity(severity_id: i64) -> i64 {
    match severity_id {
        0 => 0,  // Unknown
        1 => 1,  // Informational
        2 => 3,  // Low
        3 => 5,  // Medium
        4 => 7,  // High
        5 => 9,  // Critical
        6 => 10, // Fatal
        _ => 0,
    }
}

/// Returns the display name for a reused OCSF class UID, or `None` if unknown.
pub fn class_uid_to_name(class_uid: i64) -> Option<&'static str> {
    let name = match class_uid {
        2002 => "Vulnerability Finding",
        2003 => "Compliance Finding",
        2004 => "Detection Finding",
        3001 => "Account Change",
        3002 => "Authentication",
        3003 => "Authorize Session",
        3004 => "Entity Management",
        3005 => "User Access Management",
        5001 => "Inventory Info",
        6001 => "Web Resources Activity",
        6002 => "Application Lifecycle",
        6003 => "API Activity",
        6005 => "Datastore Activity",
        _ => return None,
    };
    Some(name)
}

/// Escapes special characters in a CEF extension value (`\ | = \n \r`).
pub fn sanitize_cef_value(value: &str) -> String {
    value
        .replace('\\', "\\\\")
        .replace('|', "\\|")
        .replace('=', "\\=")
        .replace('\n', "\\n")
        .replace('\r', "\\r")
}

/// Escapes special characters in a CEF header field (`\ |`).
pub fn sanitize_cef_header(value: &str) -> String {
    value.replace('\\', "\\\\").replace('|', "\\|")
}

fn get_int(m: &Value, key: &str, default: i64) -> i64 {
    match m.get(key) {
        Some(Value::Number(n)) => n.as_i64().or_else(|| n.as_f64().map(|f| f as i64)).unwrap_or(default),
        _ => default,
    }
}

fn get_string(m: &Value, key: &str, default: &str) -> String {
    match m.get(key) {
        Some(Value::String(s)) => s.clone(),
        Some(Value::Null) | None => default.to_string(),
        Some(other) => other.to_string(),
    }
}

/// Renders an OCSF event (as a JSON object) to a CEF syslog message string.
fn ocsf_value_to_cef(event: &Value, vendor: &str, product: &str, version: &str) -> String {
    let class_uid = get_int(event, "class_uid", 0);
    let activity_id = get_int(event, "activity_id", 0);
    let type_uid = get_int(event, "type_uid", class_uid * 100 + activity_id);
    let severity_id = get_int(event, "severity_id", 1);

    let signature_id = type_uid.to_string();
    let name = class_uid_to_name(class_uid)
        .map(|s| s.to_string())
        .unwrap_or_else(|| format!("OCSF-{class_uid}"));
    let cef_severity = ocsf_to_cef_severity(severity_id);

    let mut ext: Vec<String> = Vec::new();

    // Timestamp
    let event_time = get_string(event, "time", "");
    ext.push(format!("rt={}", sanitize_cef_value(&event_time)));

    // Message
    let msg = get_string(event, "message", "");
    if !msg.is_empty() {
        ext.push(format!("msg={}", sanitize_cef_value(&msg)));
    }

    // OCSF identifiers
    ext.push(format!("cs1={class_uid}"));
    ext.push("cs1Label=ocsf_class_uid".to_string());
    ext.push(format!("cs2={activity_id}"));
    ext.push("cs2Label=ocsf_activity_id".to_string());
    ext.push(format!("cs3={}", get_int(event, "category_uid", 6)));
    ext.push("cs3Label=ocsf_category_uid".to_string());

    // Model information
    if let Some(model_info) = event.get("model").filter(|v| v.is_object()) {
        let model_id = get_string(model_info, "model_id", "");
        if !model_id.is_empty() {
            ext.push(format!("cs4={}", sanitize_cef_value(&model_id)));
            ext.push("cs4Label=ai_model_id".to_string());
        }
        let provider = get_string(model_info, "provider", "");
        if !provider.is_empty() {
            ext.push(format!("cs5={}", sanitize_cef_value(&provider)));
            ext.push("cs5Label=ai_provider".to_string());
        }
    }

    // Agent name
    let agent_name = get_string(event, "agent_name", "");
    if !agent_name.is_empty() {
        ext.push(format!("suser={}", sanitize_cef_value(&agent_name)));
    }

    // Tool name
    let tool_name = get_string(event, "tool_name", "");
    if !tool_name.is_empty() {
        ext.push(format!("cs6={}", sanitize_cef_value(&tool_name)));
        ext.push("cs6Label=ai_tool_name".to_string());
    }

    // Security finding
    if let Some(finding) = event.get("finding").filter(|v| v.is_object()) {
        let ft = get_string(finding, "finding_type", "");
        if !ft.is_empty() {
            ext.push(format!("cat={}", sanitize_cef_value(&ft)));
        }
        if let Some(rs) = finding.get("risk_score") {
            ext.push(format!("cn1={}", json_scalar(rs)));
            ext.push("cn1Label=risk_score".to_string());
        }
        let owasp = get_string(finding, "owasp_category", "");
        if !owasp.is_empty() {
            ext.push(format!("flexString1={}", sanitize_cef_value(&owasp)));
            ext.push("flexString1Label=owasp_category".to_string());
        }
    }

    // Token usage
    if let Some(usage) = event.get("usage").filter(|v| v.is_object()) {
        if let Some(it) = usage.get("input_tokens") {
            ext.push(format!("cn2={}", json_scalar(it)));
            ext.push("cn2Label=input_tokens".to_string());
        }
        if let Some(ot) = usage.get("output_tokens") {
            ext.push(format!("cn3={}", json_scalar(ot)));
            ext.push("cn3Label=output_tokens".to_string());
        }
    }

    // Cost
    if let Some(cost) = event.get("cost").filter(|v| v.is_object()) {
        if let Some(tc) = cost.get("total_cost_usd") {
            ext.push(format!("cfp1={}", json_scalar(tc)));
            ext.push("cfp1Label=total_cost_usd".to_string());
        }
    }

    let extension_str = ext.join(" ");

    format!(
        "CEF:0|{}|{}|{}|{}|{}|{}|{}",
        sanitize_cef_header(vendor),
        sanitize_cef_header(product),
        sanitize_cef_header(version),
        signature_id,
        sanitize_cef_header(&name),
        cef_severity,
        extension_str,
    )
}

/// Renders a JSON scalar the way Go's `fmt.Sprintf("%v")` would (no quotes).
fn json_scalar(v: &Value) -> String {
    match v {
        Value::String(s) => s.clone(),
        Value::Null => "<nil>".to_string(),
        other => other.to_string(),
    }
}

/// Exports [`AIBaseEvent`] values as CEF syslog message strings.
#[derive(Debug, Clone)]
pub struct CefSyslogExporter {
    vendor: String,
    product: String,
    version: String,
}

impl Default for CefSyslogExporter {
    fn default() -> Self {
        CefSyslogExporter {
            vendor: "AITF".to_string(),
            product: "AI-Telemetry-Framework".to_string(),
            version: "0.4.0".to_string(),
        }
    }
}

impl CefSyslogExporter {
    /// Creates an exporter with AITF default header fields.
    pub fn new() -> Self {
        Self::default()
    }

    /// Sets the CEF DeviceVendor / DeviceProduct / DeviceVersion header fields.
    pub fn with_device(
        vendor: impl Into<String>,
        product: impl Into<String>,
        version: impl Into<String>,
    ) -> Self {
        CefSyslogExporter {
            vendor: vendor.into(),
            product: product.into(),
            version: version.into(),
        }
    }

    /// Converts a single event to a CEF syslog message string.
    pub fn to_cef(&self, event: &AIBaseEvent) -> String {
        let value = serde_json::to_value(event).unwrap_or(Value::Null);
        ocsf_value_to_cef(&value, &self.vendor, &self.product, &self.version)
    }

    /// Converts a batch of events to CEF syslog message strings.
    pub fn to_cef_batch(&self, events: &[AIBaseEvent]) -> Vec<String> {
        events.iter().map(|e| self.to_cef(e)).collect()
    }
}

File: telemetry/aitf/sdk/go/exporters/cef_syslog.go
// Package exporters provides AITF span exporters for Go.
//
// CEFSyslogExporter converts OTel spans to CEF (Common Event Format)
// syslog messages and sends them to any SIEM that supports CEF ingestion,
// including vendors that do not support OCSF natively (ArcSight, QRadar,
// LogRhythm, Elastic Security, etc.).
package exporters

import (
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"log"
	"net"
	"strings"
	"sync"
	"time"

	"github.com/girdav01/AITF/sdk/go/ocsf"

	sdktrace "go.opentelemetry.io/otel/sdk/trace"
)

// CEF severity mapping from OCSF severity_id.
var ocsfToCEFSeverity = map[int]int{
	0: 0,  // Unknown
	1: 1,  // Informational
	2: 3,  // Low
	3: 5,  // Medium
	4: 7,  // High
	5: 9,  // Critical
	6: 10, // Fatal
}

// classUIDToName maps the reused OCSF class UIDs that AITF emits to display
// names. Inference and tool execution both reuse API Activity (6003); they are
// distinguished by activity_id and the ai_operation profile.
var classUIDToName = map[int]string{
	2002: "Vulnerability Finding",
	2003: "Compliance Finding",
	2004: "Detection Finding",
	3002: "Authentication",
	3003: "Authorize Session",
	5001: "Inventory Info",
	6002: "Application Lifecycle",
	6003: "API Activity",
	6005: "Datastore Activity",
}

// sanitizeCEFValue escapes special characters in CEF extension values.
func sanitizeCEFValue(value string) string {
	value = strings.ReplaceAll(value, `\`, `\\`)
	value = strings.ReplaceAll(value, `|`, `\|`)
	value = strings.ReplaceAll(value, `=`, `\=`)
	value = strings.ReplaceAll(value, "\n", `\n`)
	value = strings.ReplaceAll(value, "\r", `\r`)
	return value
}

// sanitizeCEFHeader escapes special characters in CEF header fields.
func sanitizeCEFHeader(value string) string {
	value = strings.ReplaceAll(value, `\`, `\\`)
	value = strings.ReplaceAll(value, `|`, `\|`)
	return value
}

// OCSFEventToCEF converts an OCSF event map to a CEF syslog message string.
func OCSFEventToCEF(event map[string]interface{}, vendor, product, version string) string {
	classUID := getInt(event, "class_uid", 0)
	activityID := getInt(event, "activity_id", 0)
	typeUID := getInt(event, "type_uid", classUID*100+activityID)
	severityID := getInt(event, "severity_id", 1)

	signatureID := fmt.Sprintf("%d", typeUID)
	name, ok := classUIDToName[classUID]
	if !ok {
		name = fmt.Sprintf("OCSF-%d", classUID)
	}
	cefSeverity := ocsfToCEFSeverity[severityID]

	var ext []string

	// Timestamp
	eventTime := getString(event, "time", time.Now().UTC().Format(time.RFC3339))
	ext = append(ext, fmt.Sprintf("rt=%s", sanitizeCEFValue(eventTime)))

	// Message
	if msg := getString(event, "message", ""); msg != "" {
		ext = append(ext, fmt.Sprintf("msg=%s", sanitizeCEFValue(msg)))
	}

	// OCSF identifiers
	ext = append(ext, fmt.Sprintf("cs1=%d", classUID))
	ext = append(ext, "cs1Label=ocsf_class_uid")
	ext = append(ext, fmt.Sprintf("cs2=%d", activityID))
	ext = append(ext, "cs2Label=ocsf_activity_id")
	ext = append(ext, fmt.Sprintf("cs3=%d", getInt(event, "category_uid", 6)))
	ext = append(ext, "cs3Label=ocsf_category_uid")

	// Model information
	if modelInfo, ok := event["model"].(map[string]interface{}); ok {
		if modelID := getString(modelInfo, "model_id", ""); modelID != "" {
			ext = append(ext, fmt.Sprintf("cs4=%s", sanitizeCEFValue(modelID)))
			ext = append(ext, "cs4Label=ai_model_id")
		}
		if provider := getString(modelInfo, "provider", ""); provider != "" {
			ext = append(ext, fmt.Sprintf("cs5=%s", sanitizeCEFValue(provider)))
			ext = append(ext, "cs5Label=ai_provider")
		}
	}

	// Agent name
	if agentName := getString(event, "agent_name", ""); agentName != "" {
		ext = append(ext, fmt.Sprintf("suser=%s", sanitizeCEFValue(agentName)))
	}

	// Tool name
	if toolName := getString(event, "tool_name", ""); toolName != "" {
		ext = append(ext, fmt.Sprintf("cs6=%s", sanitizeCEFValue(toolName)))
		ext = append(ext, "cs6Label=ai_tool_name")
	}

	// Security finding
	if finding, ok := event["finding"].(map[string]interface{}); ok {
		if ft := getString(finding, "finding_type", ""); ft != "" {
			ext = append(ext, fmt.Sprintf("cat=%s", sanitizeCEFValue(ft)))
		}
		if rs, ok := finding["risk_score"]; ok {
			ext = append(ext, fmt.Sprintf("cn1=%v", rs))
			ext = append(ext, "cn1Label=risk_score")
		}
		if owasp := getString(finding, "owasp_category", ""); owasp != "" {
			ext = append(ext, fmt.Sprintf("flexString1=%s", sanitizeCEFValue(owasp)))
			ext = append(ext, "flexString1Label=owasp_category")
		}
	}

	// Token usage
	if usage, ok := event["usage"].(map[string]interface{}); ok {
		if it, ok := usage["input_tokens"]; ok {
			ext = append(ext, fmt.Sprintf("cn2=%v", it))
			ext = append(ext, "cn2Label=input_tokens")
		}
		if ot, ok := usage["output_tokens"]; ok {
			ext = append(ext, fmt.Sprintf("cn3=%v", ot))
			ext = append(ext, "cn3Label=output_tokens")
		}
	}

	// Cost
	if cost, ok := event["cost"].(map[string]interface{}); ok {
		if tc, ok := cost["total_cost_usd"]; ok {
			ext = append(ext, fmt.Sprintf("cfp1=%v", tc))
			ext = append(ext, "cfp1Label=total_cost_usd")
		}
	}

	extensionStr := strings.Join(ext, " ")

	return fmt.Sprintf("CEF:0|%s|%s|%s|%s|%s|%d|%s",
		sanitizeCEFHeader(vendor),
		sanitizeCEFHeader(product),
		sanitizeCEFHeader(version),
		signatureID,
		sanitizeCEFHeader(name),
		cefSeverity,
		extensionStr,
	)
}

// CEFSyslogExporter exports OTel spans as CEF syslog messages.
type CEFSyslogExporter struct {
	host          string
	port          int
	protocol      string
	useTLS        bool
	tlsConfig     *tls.Config
	vendor        string
	product       string
	version       string
	batchSize     int
	mapper        *ocsf.OCSFMapper
	conn          net.Conn
	connected     bool
	totalExported int
	mu            sync.Mutex
}

// CEFSyslogOption configures the CEFSyslogExporter.
type CEFSyslogOption func(*CEFSyslogExporter)

// WithCEFHost sets the syslog receiver host.
func WithCEFHost(host string) CEFSyslogOption {
	return func(e *CEFSyslogExporter) { e.host = host }
}

// WithCEFPort sets the syslog receiver port.
func WithCEFPort(port int) CEFSyslogOption {
	return func(e *CEFSyslogExporter) { e.port = port }
}

// WithCEFProtocol sets the transport protocol ("tcp" or "udp").
func WithCEFProtocol(protocol string) CEFSyslogOption {
	return func(e *CEFSyslogExporter) { e.protocol = strings.ToLower(protocol) }
}

// WithCEFTLS enables TLS for TCP connections.
func WithCEFTLS(enabled bool) CEFSyslogOption {
	return func(e *CEFSyslogExporter) { e.useTLS = enabled }
}

// WithCEFVendor sets the CEF DeviceVendor header field.
func WithCEFVendor(vendor string) CEFSyslogOption {
	return func(e *CEFSyslogExporter) { e.vendor = vendor }
}

// WithCEFProduct sets the CEF DeviceProduct header field.
func WithCEFProduct(product string) CEFSyslogOption {
	return func(e *CEFSyslogExporter) { e.product = product }
}

// NewCEFSyslogExporter creates a new CEF syslog exporter.
func NewCEFSyslogExporter(opts ...CEFSyslogOption) (*CEFSyslogExporter, error) {
	e := &CEFSyslogExporter{
		host:     "localhost",
		port:     514,
		protocol: "tcp",
		useTLS:   true,
		vendor:   "AITF",
		product:  "AI-Telemetry-Framework",
		version:  "0.4.0",
		mapper:   ocsf.NewOCSFMapper(),
	}
	for _, opt := range opts {
		opt(e)
	}

	if e.useTLS {
		e.tlsConfig = &tls.Config{
			MinVersion: tls.VersionTLS12,
		}
	}

	return e, nil
}

func (e *CEFSyslogExporter) connect() error {
	addr := fmt.Sprintf("%s:%d", e.host, e.port)

	switch e.protocol {
	case "tcp":
		if e.useTLS {
			conn, err := tls.DialWithDialer(
				&net.Dialer{Timeout: 10 * time.Second},
				"tcp", addr, e.tlsConfig,
			)
			if err != nil {
				return fmt.Errorf("TLS connect to %s failed: %w", addr, err)
			}
			e.conn = conn
		} else {
			conn, err := net.DialTimeout("tcp", addr, 10*time.Second)
			if err != nil {
				return fmt.Errorf("TCP connect to %s failed: %w", addr, err)
			}
			e.conn = conn
		}
	case "udp":
		conn, err := net.DialTimeout("udp", addr, 10*time.Second)
		if err != nil {
			return fmt.Errorf("UDP dial to %s failed: %w", addr, err)
		}
		e.conn = conn
	default:
		return fmt.Errorf("unsupported protocol: %q", e.protocol)
	}

	e.connected = true
	log.Printf("cef_syslog: connected to %s via %s%s",
		addr, strings.ToUpper(e.protocol),
		map[bool]string{true: "+TLS", false: ""}[e.useTLS && e.protocol == "tcp"],
	)
	return nil
}

func (e *CEFSyslogExporter) send(msg string) error {
	if !e.connected || e.conn == nil {
		if err := e.connect(); err != nil {
			return err
		}
	}

	data := []byte(msg)

	if e.protocol == "tcp" {
		// RFC 5425 octet-counting framing
		framed := fmt.Sprintf("%d %s", len(data), msg)
		_, err := e.conn.Write([]byte(framed))
		return err
	}
	// UDP: each datagram is one message
	_, err := e.conn.Write(data)
	return err
}

// ExportSpans converts spans to CEF and sends via syslog.
func (e *CEFSyslogExporter) ExportSpans(ctx context.Context, spans []sdktrace.ReadOnlySpan) error {
	var messages []string

	for _, span := range spans {
		ocsfEvent := e.mapper.MapSpan(span)
		if ocsfEvent == nil {
			continue
		}

		eventBytes, err := json.Marshal(ocsfEvent)
		if err != nil {
			continue
		}

		var eventMap map[string]interface{}
		if err := json.Unmarshal(eventBytes, &eventMap); err != nil {
			continue
		}

		cefMsg := OCSFEventToCEF(eventMap, e.vendor, e.product, e.version)
		messages = append(messages, cefMsg)
	}

	if len(messages) == 0 {
		return nil
	}

	e.mu.Lock()
	defer e.mu.Unlock()

	for _, msg := range messages {
		if err := e.send(msg); err != nil {
			// Try reconnect once
			e.connected = false
			if err := e.connect(); err != nil {
				return fmt.Errorf("cef_syslog: send failed: %w", err)
			}
			if err := e.send(msg); err != nil {
				return fmt.Errorf("cef_syslog: send retry failed: %w", err)
			}
		}
		e.totalExported++
	}

	return nil
}

// Shutdown closes the syslog connection.
func (e *CEFSyslogExporter) Shutdown(ctx context.Context) error {
	e.mu.Lock()
	defer e.mu.Unlock()
	if e.conn != nil {
		return e.conn.Close()
	}
	return nil
}

// TotalExported returns the number of CEF messages sent.
func (e *CEFSyslogExporter) TotalExported() int {
	e.mu.Lock()
	defer e.mu.Unlock()
	return e.totalExported
}

// helper to extract string from map
func getString(m map[string]interface{}, key, defaultVal string) string {
	if v, ok := m[key]; ok {
		if s, ok := v.(string); ok {
			return s
		}
		return fmt.Sprintf("%v", v)
	}
	return defaultVal
}

// helper to extract int from map
func getInt(m map[string]interface{}, key string, defaultVal int) int {
	if v, ok := m[key]; ok {
		switch n := v.(type) {
		case float64:
			return int(n)
		case int:
			return n
		case int64:
			return int(n)
		}
	}
	return defaultVal
}

// Compile-time check that CEFSyslogExporter implements SpanExporter.
var _ sdktrace.SpanExporter = (*CEFSyslogExporter)(nil)
